Fix crash when imshow encounters longdouble data

br-Zhang opened the pull request:
## PR Summary
This PR solves #10342. Support for longdouble was added in #8447, however attempting to draw anything causes a crash. This PR adds a cast to float64 when imshow encounters longdouble data.

## PR Checklist

- [x] Has Pytest style unit tests
- [x] Code is PEP 8 compliant
- [x] New features are documented, with examples if plot related
- [ ] Documentation is sphinx and numpydoc compliant
- [ ] Added an entry to doc/users/next_whats_new/ if major new feature (follow instructions in README.rst there)
- [ ] Documented in doc/api/api_changes.rst if API changed in a backward-incompatible way

anntzer commented:
So if I understand correctly, #8447 is essentially pointless?  #8447 added downcasting of float128 to float64 at normalization time, but we always run interpolation first and that step already needs to work on float64?

br-Zhang commented:
I believe that is correct.

jklymak commented:
If we just down-cast to float64, that's not *really* supporting float128, its just not crashing.  I think that float128 should error, maybe w/ a better error, and tell users they need to cast themselves, given that opinions vary wildly as to how down-casting should happen.   ping @efiring who had an opinion about all this when I added float128 before.

efiring commented:
Maybe downcast with a warning. Effectively, as a plotting library, we are *always* downcasting drastically, since plots have very limited combinations of dynamic range and resolution.  I presume the use of float128 is very rare, and I don't see any way that "supporting" it in matplotlib is going to help anyone do anything useful.  Use of float128 *should* be very rare because it is wildly platform-dependent, and doesn't even provide the advertised 128 bits.  It is not included in the numpy table of supported types:
https://docs.scipy.org/doc/numpy/user/basics.types.html
The situation is explained at the bottom of that section:
https://docs.scipy.org/doc/numpy/user/basics.types.html#extended-precision
Note that there is also a float96 which should be handled in the same (minimal) way as float128.

jklymak commented:
I'd be fine with downcast-with-warning...  For this PR, you'd want to add the warning, and an assert that its emited in the test...

anntzer commented:
Probably also worth (effectively) reverting #8447 then, given that by the time we reach colormapping, the data is certainly not float128 anymore? (dunno)

br-Zhang commented:
I've updated the pull request to add a warning before casting, remove the redundant check for casting in colors.py, and added an assert that the warning is emitted in the test.

jklymak commented on lib/matplotlib/image.py:
```diff
@@ -361,6 +362,12 @@ def _make_image(self, A, in_bbox, out_bbox, clip_bbox, magnification=1.0,
                 if inp_dtype.kind == 'f':
                     scaled_dtype = A.dtype
+                    # Cast to float64
+                    if A.dtype == np.longdouble:
```
OK, still not sure what to do about `np.float128` and `np.float96`.  Should these be trapped here as well?

br-Zhang replied:
Based on the platform version, np.float128 or np.float96 may or may not exist so adding traps for them might not work. Instead, I think casting anything that isn't float32 or float16 to float64 would be a better idea.

jklymak replied:
Really, they aren't even defined?  Yuck....

br-Zhang replied:
Yes. np.longdouble is platform dependent also. That's why I most likely have to remove the checking for warnings from the test case since the CI runs on Windows which has np.longdouble == 'float64' == 'float'. This causes either excessive warnings which will break all sorts of CI tests (if I choose to show warnings for each cast) or none at all (if I only show warnings on casts from a larger size to smaller).

jklymak commented:
Given the issues with longdouble float128 etc maybe the right thing to do is let this error?

br-Zhang commented:
The problem with longdouble is mostly with testing. The changes in this PR makes a downcast to float64 whenever it encounters longdouble, float96 or float128, which is working fine. The issue is that, on Windows, longdouble == float64 so a trap/error for longdouble would cause an error on float/float64 too. Adding a check for float64 solves this but means that warnings will only be emitted on platforms where longdouble == float128/float96, which makes it hard to write a test that asserts for a warning.

QuLogic commented on lib/matplotlib/image.py:
```diff
@@ -361,6 +362,13 @@ def _make_image(self, A, in_bbox, out_bbox, clip_bbox, magnification=1.0,
                 if inp_dtype.kind == 'f':
                     scaled_dtype = A.dtype
+                    # Cast to float64
+                    if not (A.dtype == np.float32 or A.dtype == np.float16):
```
`if A.dtype not in (np.float32, np.float16):`

br-Zhang replied:
Done.

QuLogic commented on lib/matplotlib/image.py:
```diff
@@ -361,6 +362,13 @@ def _make_image(self, A, in_bbox, out_bbox, clip_bbox, magnification=1.0,
                     scaled_dtype = A.dtype
+                    # Cast to float64
+                    if not (A.dtype == np.float32 or A.dtype == np.float16):
+                        if (A.dtype != np.float64):
```
Remove unnecessary parentheses.

br-Zhang replied:
Done.

jklymak commented on lib/matplotlib/image.py:
```diff
@@ -361,6 +362,13 @@ def _make_image(self, A, in_bbox, out_bbox, clip_bbox, magnification=1.0,
                     scaled_dtype = A.dtype
+                    # Cast to float64
+                    if A.dtype not in (np.float32, np.float16):
+                        if A.dtype != np.float64:
```
might as well just make this one if-statement....

QuLogic replied:
There are things happening after this one but not in it...

jklymak replied:
```python
if A.dtype not in (np.float32, np.float16, np.float64):
   warnings.warn()
```
should do it, right?

br-Zhang replied:
I actually did this in commit a453ba4. This led to problems in some Windows environments and failures in CI. It seems like scaled_dtype needs to be set to np.float64 instead of np.longdouble even if np.longdouble == np.float64.

jklymak replied:
`not in` returns something different than `!=`?  I'm not following the logic here at all.

br-Zhang replied:
I see now. Yes, you are correct. I forgot that in (...) is not using an equality check. This should make things much simpler.

br-Zhang replied:
The extra if statement is because on Windows environment, np.doublelong == float64 but np.longdouble is not float64. We need to make dtype to be actually float64 and not just dtype == float64 to return True.

jklymak replied:
Gotchya, so https://github.com/matplotlib/matplotlib/commit/a453ba4519d66e3ba2dd8bab3d64294ba34258cc failed because `np.doublelong == float64 ` but this should work because `np.doublelong is not in (np.float64)` .  How confusing!  Whatever you decide, this definitely needs a comment so its clear to future readers what is going on.  Sounds like you have it sorted out, but its pretty arcane.

QuLogic replied:
I think what you're missing is the [next line](https://github.com/matplotlib/matplotlib/pull/10768/files#diff-d7067d23b8918c423e30d4a3c6e13a60R371). The difference has nothing to do with `==` vs `in`, but that `scaled_dtype` is correctly updated (which causes some problem later).

jklymak replied:
If `A.dtype` is `in (np.float64)` then `scaled_dtype` is already `np.float64` per line 364 above, I'm 95% sure.

QuLogic replied:
It could be `np.longdouble` as that also equals `np.float64` (on some systems). Why this causes a problem later, I don't know...

br-Zhang replied:
I believe somewhere later, it checks that the `scaled_dtype` is `np.float32`, `np.float64`, etc. instead of ==. That's why `scaled_dtype` needs to be set to `np.float64`.

jklymak replied:
@br-Zhang I think @QuLogic is correct.  On my machine `np.longdouble` is `np.float128` and:

```
In [7]: print(np.longdouble == np.float128)
True

In [8]: print(np.longdouble in (np.float64, np.float32, np.float128))
True
```

so I don't understand at all whats going on.

br-Zhang replied:
It's very platform specific. Running on Win-64 gives:
`
>>> numpy.longdouble in (numpy.float16, numpy.float64)
False
`

jklymak replied:
But `numpy.longdouble == numpy.float64` returns true?

br-Zhang replied:
Very weird.
``` 
>>> numpy.longdouble == numpy.float64
True
```

QuLogic replied:
Again, the problem doesn't have anything to do with the check here; it's the use of `scaled_dtype` later that causes a problem. If you flatten the condition, then `scaled_dtype=np.longdouble` is not replaced by `np.float64` (on Windows only!) It's then crashing in C code later (which is arguably a NumPy inconsistency) which doesn't understand `np.longdouble`.

jklymak replied:
OK; I get it now.  everything not `np.float16/32` has to be explicitly set to `np.float64`, even if python thinks `np.longdouble` is equivalent to `np.float64`, because the C code doesn't think its equivalent.  

Sorry to be so dense, but I do think its pretty bad that the C code ends up incompatible in this way.

QuLogic commented on lib/matplotlib/image.py:
```diff
@@ -10,6 +10,7 @@
 from io import BytesIO
 
 from math import ceil
+import warnings
```
This belongs after `logging`.

br-Zhang replied:
Done

br-Zhang commented:
Squashed all the previous commits into one commit.

jklymak commented:
I'll merge tomorrow if no one does so earlier.  Just want to leave it up if anyone else wants to chime in.  But I think I understand what is going on now and this fix seems fine.
